Core/Add campaign DB2 support and quest campaign handle to npc

dio85 opened the pull request:
Introduces new DB2 structures, storage, and SQL tables for campaign and campaign quest lines. Adds logic to detect campaign quests and updates quest dialog status handling for campaign quests in Player.cpp. This enables campaign-specific quest behaviors and data access throughout the server codebase.

Now check the conditions, if campaign active, that quests showed with campaign icon.

**Changes proposed:**

-  Added Campaign Icon handle to the npc
-  Added Future campaign icon handle
-  Added trivial campaign icon handle
-  Added Conditions check.


**Tests performed:**
Win11
(Does it build, tested in-game, etc.)

dio85 committed:
Core/Add campaign DB2 support and quest campaign detection

Introduces new DB2 structures, storage, and SQL tables for campaign and campaign quest lines. Adds logic to detect campaign quests and updates quest dialog status handling for campaign quests in Player.cpp. This enables campaign-specific quest behaviors and data access throughout the server codebase.

meji46 commented on src/server/game/DataStores/DB2Stores.cpp:
```diff
@@ -516,6 +520,7 @@ namespace
     std::unordered_map<uint32, uint8> _pvpItemBonus;
     PvpTalentSlotUnlockEntry const* _pvpTalentSlotUnlock[MAX_PVP_TALENT_SLOTS];
     std::unordered_map<uint32, std::vector<QuestLineXQuestEntry const*>> _questsByQuestLine;
+    std::unordered_map<uint32, std::vector<QuestLineXQuestEntry const*>> _questsByQuestXLine;
```
I think _questXLinesByQuest is a better name (you store all QuestLineXQuestEntry for each quest)

meji46 commented on src/server/game/DataStores/DB2Stores.cpp:
```diff
@@ -1510,6 +1520,14 @@ void DB2Manager::IndexLoadedStores()
     }
 
+    {
+        for (QuestLineXQuestEntry const* entry : sQuestLineXQuestStore)
```
Instead of creating a new for loop, reuse the one above

meji46 commented on src/server/game/DataStores/DB2Stores.cpp:
```diff
@@ -2115,6 +2133,33 @@ int32 const* DB2Manager::GetBroadcastTextDuration(uint32 broadcastTextId, Locale
 
+QuestLineXQuestEntry const* DB2Manager::GetQuestLineXQuestForQuest(uint32 questId) const
+{
+    for (auto const& [questLineId, quests] : _questsByQuestXLine)
```
Use Trinity::Containers::MapGetValuePtr

meji46 commented on src/server/game/DataStores/DB2Stores.cpp:
```diff
@@ -2115,6 +2133,33 @@ int32 const* DB2Manager::GetBroadcastTextDuration(uint32 broadcastTextId, Locale
+    for (auto const& [questLineId, quests] : _questsByQuestXLine)
+    {
+        for (QuestLineXQuestEntry const* questLineXQuest : quests)
+            if (questLineXQuest->QuestID == questId)
```
No need to check that.

In addition, you only return the first result and you have cases like quest 48111 that has two questlines (and only one is linked to a campaign). So depending on which one is returned, the quest could be considered as part of a campaign or not (you need to check if the player can have that campaign active)

meji46 commented on src/server/game/DataStores/DB2Stores.cpp:
```diff
@@ -2115,6 +2133,33 @@ int32 const* DB2Manager::GetBroadcastTextDuration(uint32 broadcastTextId, Locale
+
+CampaignXQuestLineEntry const* DB2Manager::GetCampaignForQuestLine(uint32 questLineId) const
+{
+    auto itr = _campaignByQuestLine.find(questLineId);
```
Trinity::Containers::MapGetValuePtr

meji46 commented on src/server/game/DataStores/DB2Stores.cpp:
```diff
@@ -2115,6 +2133,33 @@ int32 const* DB2Manager::GetBroadcastTextDuration(uint32 broadcastTextId, Locale
+
+CampaignEntry const* DB2Manager::GetCampaign(uint32 campaignId) const
+{
+    auto itr = _campaigns.find(campaignId);
```
Trinity::Containers::MapGetValuePtr

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Core/Refactor quest and campaign lookup logic

Changed internal data structures to use vectors for quest and campaign lookups, allowing support for multiple entries per quest or quest line. Updated related methods and logic in DB2Manager and QuestDef to accommodate these changes and improve campaign detection for quests.

Shauren commented on src/server/game/DataStores/DB2Stores.cpp:
```diff
@@ -470,6 +472,8 @@ namespace
     std::unordered_map<std::pair<uint32 /*azeriteUnlockSetId*/, ItemContext>, std::array<uint8, MAX_AZERITE_EMPOWERED_TIER>> _azeriteTierUnlockLevels;
     std::unordered_map<std::pair<uint32 /*broadcastTextId*/, CascLocaleBit /*cascLocaleBit*/>, int32> _broadcastTextDurations;
+    std::unordered_map<uint32, std::vector<CampaignXQuestLineEntry const*>> _campaignsByQuestLine;
+    std::unordered_map<uint32, CampaignEntry const*> _campaigns;
```
is this used anywhere?

dio85 replied:
> is this used anywhere?
_campaigns no.

dio85 committed:
Core/Refactor campaign quest detection logic

Changed campaign quest detection to use player-specific conditions by passing Player to IsCampaign. Updated DB2Manager to return a single CampaignXQuestLineEntry per quest line and adjusted related logic for campaign lookup and sorting. Modified Player.cpp to use the new IsCampaign signature for quest dialog status checks.

dio85 committed:
Core/Remove unnecessary Player.h include from QuestDef.h

The direct inclusion of Player.h in QuestDef.h was removed, as a forward declaration of the Player class is sufficient. This helps reduce unnecessary dependencies and improves compilation times.

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Implement Chromie Time system and expansion selection

Adds Chromie Time support, including database tables, server logic, packet handling, and spell effect for expansion selection. Updates LFG, player, and NPC handler code to integrate Chromie Time expansion selection, locking, and player state management.

TODO playercondition check, and condition check at lfgmgr and the npchandler

dio85 committed:
Merge branch 'TrinityCore:master' into tww_develop

dio85 committed:
Merge branch 'TrinityCore:master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop

Shauren commented:
Just letting you know that this PR contains more than just quest campaign stuff

dio85 commented:
> Just letting you know that this PR contains more than just quest campaign stuff

whoops, i see idk how, i just reopen this pr oO

Shauren commented:
Its because you opened the PR using your `tww_develop` branch where you are pushing all your stuff (PRs are not snapshots of a branch made when you open them, they dynamically update)

dio85 committed:
Merge remote-tracking branch 'upstream/master' into tww_develop
